### PR TITLE
[ICD]Add optional common ICD key to indicate ICD operating mode (LIT/SIT)

### LINKED ISSUE #28228: [ICD] Advertise ICD discovery key when the device is operating as a LIT ICD
#### Description
To inform clients the operating mode a LIT ICD, the device needs to :

- Advertise 1 on the ICD discovery key when it is operating as a LIT ICD
- Advertise 0 on the ICD discovery key when it is operating as a SIT ICD.

A SIT ICD will not advertise this key.

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
fixes #28228

Add the new optional common Key `ICD` in the DNSSD advertisement. This key informs if the ICD device is currently operating as a long idle time (LIT) device or has a short idle time device (SIT). The ICD key is not advertised for non-ICD devices, or ICD devices that do not support LIT.

Add tests for this new key.
Also, add and fix some tests for the previously added SAT key.

### CODE REVIEW

File: src/app/server/Dnssd.cpp [hunk 1]
```diff
@@ -165,14 +165,23 @@ CHIP_ERROR DnssdServer::AdvertiseOperational()
...
+                                       .SetPeerId(fabricInfo.GetPeerId())
+                                       .SetMac(mac)
+                                       .SetPort(GetSecuredPort())
+                                       .SetInterfaceId(GetInterfaceId())
+                                       .SetLocalMRPConfig(GetLocalMRPConfig())
+                                       .SetTcpSupported(Optional<bool>(INET_CONFIG_ENABLE_TCP_ENDPOINT))
+                                       .EnableIpV4(true);
+
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+        // Only advertise the ICD key if the device can be operate as LIT
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:178]:
```suggestion
        // Only advertise the ICD key if the device can operate as a LIT.
```

File: src/app/server/Dnssd.cpp [hunk 2]
```diff
@@ -243,6 +252,15 @@ CHIP_ERROR DnssdServer::Advertise(bool commissionableNode, chip::Dnssd::Commissi
 
     advertiseParameters.SetLocalMRPConfig(GetLocalMRPConfig()).SetTcpSupported(Optional<bool>(INET_CONFIG_ENABLE_TCP_ENDPOINT));
 
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+    // Only advertise the ICD key if the device can be operate as LIT
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:256]: See above.  And maybe this should be factored out into a helper function?

File: src/lib/address_resolve/AddressResolve.h [hunk 3]
```diff
@@ -33,7 +33,8 @@ struct ResolveResult
 {
     Transport::PeerAddress address;
     ReliableMessageProtocolConfig mrpRemoteConfig;
-    bool supportsTcp = false;
+    bool supportsTcp      = false;
+    bool ICDOperatesAsLIT = false;
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve.h:37]:
```suggestion
    bool icdOperatesAsLIT = false;
```

File: src/lib/dnssd/Advertiser.h [hunk 4]
```diff
@@ -101,6 +101,13 @@ class BaseAdvertisingParams
     }
     Optional<bool> GetTcpSupported() const { return mTcpSupported; }
 
+    Derived & SetICDOperatesAsLIT(Optional<bool> ICDOperatesAsLIT)
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Advertiser.h:104]:
```suggestion
    Derived & SetICDOperatesAsLIT(Optional<bool> operatesAsLIT)
```
to match sfile style... or `icdOperatesAsLIT`, I guess.

File: src/lib/dnssd/Advertiser_ImplMinimalMdns.cpp [hunk 5]
```diff
@@ -219,6 +219,7 @@ class AdvertiserMinMdns : public ServiceAdvertiser,
         char sessionActiveThresholdBuf[KeySize(TxtFieldKey::kSessionActiveThreshold) +
                                        ValSize(TxtFieldKey::kSessionActiveThreshold) + 2];
         char tcpSupportedBuf[KeySize(TxtFieldKey::kTcpSupported) + ValSize(TxtFieldKey::kTcpSupported) + 2];
+        char ICDAsLITBuf[KeySize(TxtFieldKey::kLongIdleTimeICD) + ValSize(TxtFieldKey::kLongIdleTimeICD) + 2];
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Advertiser_ImplMinimalMdns.cpp:222]: Again, would be better to have `icd` here to match style.

File: src/lib/dnssd/Resolver.h [hunk 6]
```diff
@@ -48,6 +48,7 @@ struct CommonResolutionData
     uint16_t port                         = 0;
     char hostName[kHostNameMaxLength + 1] = {};
     bool supportsTcp                      = false;
+    Optional<bool> ICDOperatesAsLIT;
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver.h:51]: See above on naming.

File: src/lib/dnssd/Resolver.h [hunk 7]
```diff
@@ -127,7 +130,23 @@ struct CommonResolutionData
...
+            ChipLogDetail(Discovery, "\tMrp Active Threshold: %u ms", mrpRetryActiveThreshold.Value().count());
+        }
+        else
+        {
+            ChipLogDetail(Discovery, "\tMrp Active Threshold: not present");
+        }
         ChipLogDetail(Discovery, "\tTCP Supported: %d", supportsTcp);
+        if (ICDOperatesAsLIT.HasValue())
+        {
+            ChipLogDetail(Discovery, "\tThe ICD operates in %s", ICDOperatesAsLIT.Value() ? "LIT" : "SIT");
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/Resolver.h:144]:
```suggestion
            ChipLogDetail(Discovery, "\tThe ICD is operating as a %s", ICDOperatesAsLIT.Value() ? "LIT" : "SIT");
```

File: src/lib/shell/commands/Dns.cpp [hunk 8]
```diff
@@ -116,7 +118,18 @@ class DnsShellResolverDelegate : public Dnssd::CommissioningResolveDelegate, pub
...
+        if (nodeData.resolutionData.GetMrpRetryActiveThreshold().HasValue())
+        {
+            streamer_printf(streamer_get(), "   MRP retry active threshold time: %" PRIu32 "ms\r\n",
+                            nodeData.resolutionData.GetMrpRetryActiveThreshold().Value());
+        }
         streamer_printf(streamer_get(), "   Supports TCP: %s\r\n", nodeData.resolutionData.supportsTcp ? "yes" : "no");
+
+        if (nodeData.resolutionData.ICDOperatesAsLIT.HasValue())
+        {
+            streamer_printf(streamer_get(), "   ICD is operating as : %s ms\r\n",
```
Comment (bzbarsky-apple) [inline src/lib/shell/commands/Dns.cpp:130]: What's the "ms" doing here?

File: src/lib/dnssd/Advertiser.h (hunk 4 above)
Comment (jmartinez-silabs) [reply src/lib/dnssd/Advertiser.h:104]: Didn't we discuss before that ICD was an acronym and therefore should be always capital? 
I will use `operatesAsLIT` when it is clear enough that this is tied to ICD like here. Otherwise I'll move the ICD later in the variable name

Comment (bzbarsky-apple) [reply src/lib/dnssd/Advertiser.h:104]: It's an acronym, so should not be "Icd".  It should be "ICD" or "icd"; but the point is it gets capitalized or not as a unit...

File: src/app/icd/ICDManager.cpp [hunk 9]
```diff
@@ -101,7 +101,8 @@ void ICDManager::UpdateICDMode()
 
     // The Check In Protocol Feature is required and the slow polling interval shall also be greater than 15 seconds
     // to run an ICD in LIT mode.
-    if (GetSlowPollingInterval() > GetSITPollingThreshold() && SupportsCheckInProtocol())
+    if (GetSlowPollingInterval() > GetSITPollingThreshold() && SupportsFeature(Feature::kCheckInProtocolSupport) &&
```
[the remark this answers is not in the document]
Comment (jmartinez-silabs) [reply src/app/icd/ICDManager.cpp]: done

File: src/lib/dnssd/Advertiser.h (hunk 4 above)
Comment (jmartinez-silabs) [reply src/lib/dnssd/Advertiser.h]: Should always be uppercase in my opinion. Anyhow, I renamed the vars  in a way that addresses this comment

File: src/app/icd/ICDManager.cpp [hunk 10]
```diff
@@ -52,6 +52,12 @@ void ICDManager::Init(PersistentStorageDelegate * storage, FabricTable * fabricT
     VerifyOrDie(stateObserver != nullptr);
     VerifyOrDie(symmetricKeystore != nullptr);
 
+    bool supportLIT = SupportsFeature(Feature::kLongIdleTimeSupport);
+    VerifyOrDieWithMsg((supportLIT == false) || (supportLIT && SupportsFeature(Feature::kCheckInProtocolSupport)), AppServer,
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:56]:
```suggestion
    VerifyOrDieWithMsg((supportLIT == false) || SupportsFeature(Feature::kCheckInProtocolSupport), AppServer,
```
From the boolean standpoint, you don't need to check if supportLit == True since you know it is true if you are checking for the Check-In Protocol Support.

File: src/app/icd/ICDManager.cpp [hunk 11]
```diff
@@ -52,6 +52,12 @@ void ICDManager::Init(PersistentStorageDelegate * storage, FabricTable * fabricT
...
+                       "The CheckIn protocol feature is required for LIT support");
+    VerifyOrDieWithMsg((supportLIT == false) || (supportLIT && SupportsFeature(Feature::kUserActiveModeTrigger)), AppServer,
```
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:58]:
```suggestion
    VerifyOrDieWithMsg((supportLIT == false) || SupportsFeature(Feature::kUserActiveModeTrigger), AppServer,
```
same comment

File: src/app/tests/suites/commands/discovery/DiscoveryCommands.cpp [hunk 12]
```diff
@@ -175,6 +175,17 @@ void DiscoveryCommands::OnNodeDiscovered(const chip::Dnssd::DiscoveredNodeData &
         data.mrpRetryIntervalActive.SetValue(nodeData.resolutionData.mrpRetryIntervalActive.Value().count());
     }
 
+    // TODO need to add new entries for kDefaultResponse in DiscoveryCommands.js (project-chip/zap) before adding this
```
Comment (mkardous-silabs) [inline src/app/tests/suites/commands/discovery/DiscoveryCommands.cpp:178]: worth while to create an issue to track this?

File: src/lib/dnssd/Advertiser.h [hunk 13]
```diff
@@ -101,6 +101,13 @@ class BaseAdvertisingParams
     }
     Optional<bool> GetTcpSupported() const { return mTcpSupported; }
 
+    Derived & SetICDOperatingAsLIT(Optional<bool> operatesAsLIT)
+    {
+        mICDOperatesAsLIT = operatesAsLIT;
+        return *reinterpret_cast<Derived *>(this);
+    }
+    Optional<bool> GetICDOperatingAsLIT() const { return mICDOperatesAsLIT; }
```
Comment (mkardous-silabs) [inline src/lib/dnssd/Advertiser.h:109]: Just to use the same terminalogy every where, this should be renamed.

File: src/lib/dnssd/Advertiser.h [hunk 14]
```diff
@@ -109,6 +116,7 @@ class BaseAdvertisingParams
     size_t mMacLength                = 0;
     Optional<ReliableMessageProtocolConfig> mLocalMRPConfig;
     Optional<bool> mTcpSupported;
+    Optional<bool> mICDOperatesAsLIT;
```
Comment (mkardous-silabs) [inline src/lib/dnssd/Advertiser.h:119]: same comment

**Review by mkardous-silabs — APPROVED** [review]
approved with a few nits

File: src/app/tests/suites/commands/discovery/DiscoveryCommands.cpp (hunk 12 above)
Comment (jmartinez-silabs) [reply src/app/tests/suites/commands/discovery/DiscoveryCommands.cpp:178]: not really. It was already not updated and is part of the simulated cluster test. It is not really needed. I already have a PR for zap that I just need to test. I'll do it when I have some time

File: src/lib/dnssd/Advertiser.h (hunk 13 above)
Comment (jmartinez-silabs) [reply src/lib/dnssd/Advertiser.h:109]: I don't really agree. it is a separate structure and this variable used internally by the getter setter of the same name. In this case, reducing the variable name seems more important than matching another variable in another class

File: src/app/server/Dnssd.cpp [hunk 15]
```diff
@@ -145,6 +145,19 @@ CHIP_ERROR DnssdServer::SetEphemeralDiscriminator(Optional<uint16_t> discriminat
     return CHIP_NO_ERROR;
 }
 
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+template <class Derived>
+void DnssdServer::AddICDKeyToAdvertisement(Dnssd::BaseAdvertisingParams<Derived> & advParams)
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:150]: Wouldn't this be simpler as:
```suggestion
template <class AdvertisingParams>
void DnssdServer::AddICDKeyToAdvertisement(AdvertisingParams & advParams)
```
and then the callsites would not need to manually specialize the template.

File: src/app/server/Dnssd.cpp [hunk 16]
```diff
@@ -165,14 +178,18 @@ CHIP_ERROR DnssdServer::AdvertiseOperational()
...
+                                       .SetPeerId(fabricInfo.GetPeerId())
+                                       .SetMac(mac)
+                                       .SetPort(GetSecuredPort())
+                                       .SetInterfaceId(GetInterfaceId())
+                                       .SetLocalMRPConfig(GetLocalMRPConfig())
+                                       .SetTcpSupported(Optional<bool>(INET_CONFIG_ENABLE_TCP_ENDPOINT))
+                                       .EnableIpV4(true);
+
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+        AddICDKeyToAdvertisement<Dnssd::OperationalAdvertisingParameters>(advertiseParameters);
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:191]: So this would be:
```suggestion
        AddICDKeyToAdvertisement(advertiseParameters);
```

File: src/app/server/Dnssd.cpp [hunk 17]
```diff
@@ -243,6 +260,10 @@ CHIP_ERROR DnssdServer::Advertise(bool commissionableNode, chip::Dnssd::Commissi
 
     advertiseParameters.SetLocalMRPConfig(GetLocalMRPConfig()).SetTcpSupported(Optional<bool>(INET_CONFIG_ENABLE_TCP_ENDPOINT));
 
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+    AddICDKeyToAdvertisement<Dnssd::CommissionAdvertisingParameters>(advertiseParameters);
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.cpp:264]:
```suggestion
    AddICDKeyToAdvertisement(advertiseParameters);
```

File: src/app/server/Dnssd.h [hunk 18]
```diff
@@ -85,6 +85,10 @@ class DLL_EXPORT DnssdServer
     void OnExtendedDiscoveryExpiration(System::Layer * aSystemLayer, void * aAppState);
 #endif // CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY
 
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+    template <class Derived>
+    void AddICDKeyToAdvertisement(Dnssd::BaseAdvertisingParams<Derived> & advParams);
```
Comment (bzbarsky-apple) [inline src/app/server/Dnssd.h:90]: See comments above.

File: src/lib/shell/commands/Dns.cpp [hunk 19]
```diff
@@ -59,9 +59,11 @@ class DnsShellResolverDelegate : public Dnssd::CommissioningResolveDelegate, pub
                         result.mrpRemoteConfig.mIdleRetransTimeout.count());
         streamer_printf(streamer_get(), "   MRP ACTIVE retransmit timeout: %u ms\r\n",
                         result.mrpRemoteConfig.mActiveRetransTimeout.count());
-        streamer_printf(streamer_get(), "   MRP ACTIVE Threshold timet:    %u ms\r\n",
+        streamer_printf(streamer_get(), "   MRP ACTIVE Threshold time:     %u ms\r\n",
                         result.mrpRemoteConfig.mActiveThresholdTime.count());
 
+        streamer_printf(streamer_get(), "   ICD is operating as a:         %s ms\r\n", result.isICDOperatingAsLIT ? "LIT" : "SIT");
```
Comment (bzbarsky-apple) [inline src/lib/shell/commands/Dns.cpp:65]:
```suggestion
        streamer_printf(streamer_get(), "   ICD is operating as a:         %s\r\n", result.isICDOperatingAsLIT ? "LIT" : "SIT");
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/server/Dnssd.cpp (hunk 15 above)
Comment (jmartinez-silabs) [reply src/app/server/Dnssd.cpp]: Yes, simpler/cleaner ! Thanks, applied.

File: src/lib/dnssd/minimal_mdns/tests/TestResponseSender.cpp [hunk 20]
```diff
@@ -312,36 +312,40 @@ void PtrSrvTxtMultipleRespondersToInstance(nlTestSuite * inSuite, void * inConte
 
 void PtrSrvTxtMultipleRespondersToServiceListing(nlTestSuite * inSuite, void * inContext)
 {
-    CommonTestElements common1(inSuite, "test1");
-    CommonTestElements common2(inSuite, "test2");
-
+    // CommonTestElements common1(inSuite, "test1");
+    // CommonTestElements common2(inSuite, "test2");
```
Comment (andy31415) [inline src/lib/dnssd/minimal_mdns/tests/TestResponseSender.cpp:316]: Why was this changed to heap?

- should use std::unique_ptr instead of new/delete
- there should be a comment why we use heap here (since previously we did not seem to need it)
